crates/ouroboros/src/praos/header.rs

```diff
@@ -0,0 +1,428 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{
+    ed25519, issuer_to_pool_id,
+    kes::{KesPublicKey, KesSignature},
+    math::{ExpOrdering, FixedDecimal, FixedPrecision},
+    vrf, BlockHeader, Hash, Hasher, OperationalCert, PoolId, VrfCert,
+};
+use amaru_ouroboros_traits::HasStakeDistribution;
+use std::{array::TryFromSliceError, ops::Deref, sync::LazyLock};
+use thiserror::Error;
+
+/// The certified natural max value represents 2^256 in praos consensus
+#[allow(clippy::expect_used)]
+static CERTIFIED_NATURAL_MAX: LazyLock<FixedDecimal> = LazyLock::new(|| {
+    FixedDecimal::from_str(
+        "1157920892373161954235709850086879078532699846656405640394575840079131296399360000000000000000000000000000000000",
+        34,
+    )
+        .expect("Infallible")
+});
+
+// ------------------------------------------------------------------ assert_all
+
+#[derive(Error, Debug)]
+pub enum AssertHeaderError {
+    #[error("{0}")]
+    KnownLeaderVrf(#[from] AssertKnownLeaderVrfError),
+    #[error("{0}")]
+    VrfProof(#[from] AssertVrfProofError),
+    #[error("{0}")]
+    LeaderStake(#[from] AssertLeaderStakeError),
+    #[error("{0}")]
+    KesSignature(#[from] AssertKesSignatureError),
+    #[error("{0}")]
+    OperationalCertificate(#[from] AssertOperationalCertificateError),
+    #[error("{0}")]
+    TryFromSliceError(#[from] TryFromSliceError),
+    #[error("Unknown pool: {}", hex::encode(&pool[0..7]))]
+    UnknownPool { pool: PoolId },
+}
+
+pub type Assertion<'a> = Box<dyn Fn() -> Result<(), AssertHeaderError> + Send + Sync + 'a>;
+
+#[allow(clippy::result_large_err)]
+pub fn assert_all<'a>(
+    header: &'a BlockHeader<'a>,
+    ledger_state: &'a dyn HasStakeDistribution,
+    epoch_nonce: Hash<32>,
+    active_slot_coeff: &'a FixedDecimal,
+) -> Result<Vec<Assertion<'a>>, AssertHeaderError> {
+    // Grab all the values we need to validate the block
+    let absolute_slot = header.header_body.slot;
+    let issuer = ed25519::PublicKey::from(<[u8; ed25519::PublicKey::SIZE]>::try_from(
+        &header.header_body.issuer_vkey[..],
+    )?);
+    let pool: PoolId = issuer_to_pool_id(&issuer);
+
+    // TODO: Pallas should hold sized slices
+    let declared_vrf_key: &'a [u8; vrf::PublicKey::SIZE] =
+        header.header_body.vrf_vkey[..].try_into()?;
+
+    let (registered_vrf_key, leader_relative_stake): (
+        Hash<{ vrf::PublicKey::HASH_SIZE }>,
+        FixedDecimal,
+    ) = ledger_state
+        .get_pool(absolute_slot, &pool)
+        .map(|pool| {
+            (
+                pool.vrf,
+                FixedDecimal::from(pool.stake) / FixedDecimal::from(pool.active_stake),
+            )
+        })
+        .ok_or(AssertHeaderError::UnknownPool { pool })?;
+
+    Ok(vec![
+        Box::new(move || {
+            AssertKnownLeaderVrfError::new(
```

Comment:
It's a bit weird the assertions should be named `XXXError`, why not the other way round, eg. name the assertions `XXX`? Especially as some of those assertions can return different error (sub) types.